File: 15-3sum/15-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans = []
        nums = sorted(nums)
        previous = None 
        for (idx, el) in enumerate(nums):
            if (previous == el): continue
            for idx1, idx2 in two_sum(nums[idx + 1:], -el):
                ans.append([el, nums[idx1 + idx + 1], nums[idx2 + idx + 1]])
            previous = el
        return ans 
        
def two_sum(numbers, target):
    left = 0
    right = len(numbers) - 1
    ans = []
    while left < right:
        # print(left, right)
        if numbers[left] + numbers[right] == target: 
            ans.append((left, right))
            while left < right and numbers[left+1] == numbers[left]:
                left += 1
            while left < right and numbers[right-1] == numbers[right]:
                right -= 1
        if (numbers[left] + numbers[right]) > target:
            right -= 1
        else:
            left += 1

    return ans

File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    
    const compFn = (a, b) => a - b;
    const seen = new Set();
    nums.sort(compFn);
    
    console.log(nums)
    const ans = [];
    for (const [idx, target] of nums.entries()) {
        const twoSumsAns = twoSum(nums.slice(idx + 1), -target)
        // console.log("Target: " + (-target))
        // twoSumsAns.forEach(item => {
        //     console.log([item[0] + idx + 1, item[1] + idx + 1])
        // })
        for (const [_, __] of twoSumsAns) {  
            //nums[_] + nums[__] == -target, i.e. they sum to 0
            const triplet = [nums[_ + idx + 1], nums[__ + idx + 1], target];
            triplet.sort(compFn);
            const triplet_srl = JSON.stringify(triplet);
            if (!seen.has(triplet_srl)){
                ans.push(triplet);
                seen.add(triplet_srl);
            }
        }
    }
    
    return ans;
};


var twoSum = function(numbers, target) {
    var left = 0, 
        right = numbers.length - 1;

    const ans = [];
    while (left < right) {
        if (numbers[left] + numbers[right] > target) {
            right--;
        } else if (numbers[left] + numbers[right] < target) {
            left++;
        } else {
            ans.push([left, right])
            if (numbers[left + 1] + numbers[right] > target) {
                right--;
            } else {
                left++;
            }
        }
    }

    return ans;
};
